File: 3. Behavioral Design Patterns/State/Java/StateTemplate.java
// State Pattern – Java Template

interface State { void handle(Context context); }
class Context { private State state; Context(State s){ state = s; } void setState(State s){ state = s; } void request(){ state.handle(this); } }
class ConcreteStateA implements State { public void handle(Context c){ /* work */ c.setState(new ConcreteStateB()); } }
class ConcreteStateB implements State { public void handle(Context c){ /* work */ c.setState(new ConcreteStateA()); } }
public class StateTemplate { public static void main(String[] args){ Context c = new Context(new ConcreteStateA()); c.request(); c.request(); } }

File: 3. Behavioral Design Patterns/State/C#/StateTemplate.cs
// State Pattern – C# Template

namespace StateTemplate
{
    interface IState { void Handle(Context context); }

    class Context
    {
        private IState _state;
        public Context(IState initialState) { _state = initialState; }
        public void SetState(IState state) { _state = state; }
        public void Request() { _state.Handle(this); }
    }

    class ConcreteStateA : IState { public void Handle(Context context) { /* work */ context.SetState(new ConcreteStateB()); } }
    class ConcreteStateB : IState { public void Handle(Context context) { /* work */ context.SetState(new ConcreteStateA()); } }

    class Program { static void Main() { var ctx = new Context(new ConcreteStateA()); ctx.Request(); ctx.Request(); } }
}
